### ISSUE TITLE
possible bug / numerical inconsistency under torch.compile

### ISSUE THREAD

**Issue by kaixuanliu** [pytorch/pytorch#185894]
### 🐛 Describe the bug

possible bug / numerical inconsistency under torch.compile. I am not sure if it is a bug. Can you help check? Here is an example
```
import torch
import torchvision.transforms.v2.functional as F
from torchvision.transforms import InterpolationMode

device = "cuda"
new_size = (48, 48)
interpolation = InterpolationMode.BILINEAR

def resize_uint8(x):
    return F.resize(x, new_size, interpolation=interpolation, antialias=False)

compiled_resize_uint8 = torch.compile(resize_uint8, mode="reduce-overhead")

for seed in range(10):
    torch.compiler.reset()
    torch.manual_seed(seed)
    image = torch.randint(0, 255, (1, 3, 224, 224), dtype=torch.uint8, device=device)
    eager = resize_uint8(image)
    compiled = compiled_resize_uint8(image)

    diff = eager.to(torch.int16) - compiled.to(torch.int16)
    print(
        seed,
        "mismatch:",
        (diff != 0).sum().item(),
        "max_abs:",
        diff.abs().max().item(),
    )

    if (diff != 0).any():
        idx = (diff != 0).nonzero()[0].tolist()
        print("first mismatch:", idx, eager[tuple(idx)].item(), compiled[tuple(idx)].item())
```
Our expectation: torch.compile should not change the eager output of torchvision.transforms.v2.functional.resize for the same uint8 input, especially when the output is still uint8, even if the difference is only a few pixels by 1 stop, it will cause the strict equivalence test to fail.

### Versions

```
PyTorch version: 2.12.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar  3 2026, 12:15:18) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.292-1.el8.elrepo.x86_64-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe

Nvidia driver version: 570.133.20
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.20.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.20.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
BIOS Vendor ID:                     Intel(R) Corporation
Model name:                         Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
BIOS Model name:                    Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz  CPU @ 2.6GHz
BIOS CPU family:                    179
CPU family:                         6
Model:                              106
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           6
Frequency boost:                    enabled
CPU(s) scaling MHz:                 97%
CPU max MHz:                        3400.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           80 MiB (64 instances)
L3 cache:                           96 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Vulnerable: No microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] apollo-torch==1.0.3
[pip3] galore-torch==1.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.20.1
[pip3] onnx-ir==0.2.0
[pip3] onnxscript==0.6.2
[pip3] torch==2.12.0+cu126
[pip3] torchdata==0.11.0
[pip3] torchvision==0.27.0+cu126
[pip3] transformer_engine_torch==2.13.0
[pip3] triton==3.7.0
```

cc @ezyang @gchanan @kadeng @msaroufim @datumbox @vfdev-5 @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste

**Comment by vfdev-5** [comment]
@kaixuanliu it is a rounding problem.
For example, uint8 case gives:
```python
import torch
import torchvision.transforms.v2.functional as F
from torchvision.transforms import InterpolationMode

device = "cuda"
new_size = (48, 48)
interpolation = InterpolationMode.BILINEAR

def resize_uint8(x):
    return F.resize(x, new_size, interpolation=interpolation, antialias=False)

compiled_resize_uint8 = torch.compile(resize_uint8, mode="reduce-overhead")

for seed in range(1):
    torch.compiler.reset()
    torch.manual_seed(seed)
    image = torch.randint(0, 255, (1, 3, 224, 224), dtype=torch.uint8, device=device)
    eager = resize_uint8(image)
    compiled = compiled_resize_uint8(image)

    diff = eager.to(torch.float32) - compiled.to(torch.float32)
    print(
        seed,
        "mismatch:",
        (diff != 0).sum().item(),
        "max_abs:",
        diff.abs().max().item(),
    )

    if (diff != 0).any():
        idx = (diff != 0).nonzero()[0].tolist()
        print("first mismatch:", idx, eager[tuple(idx)].item(), compiled[tuple(idx)].item())
```
Output:
```
0 mismatch: 3 max_abs: 1.0
first mismatch: [0, 1, 36, 25] 234 235
```
and if we execute the same in float32 and print the output of `eager[0, 1, 36, 25], compiled[0, 1, 36, 25]` we'll see the following:
```python
import torch
import torchvision.transforms.v2.functional as F
from torchvision.transforms import InterpolationMode

device = "cuda"
new_size = (48, 48)
interpolation = InterpolationMode.BILINEAR

def resize_uint8(x):
    return F.resize(x, new_size, interpolation=interpolation, antialias=False)

compiled_resize_uint8 = torch.compile(resize_uint8, mode="reduce-overhead")

for seed in range(1):
    torch.compiler.reset()
    torch.manual_seed(seed)
    image = torch.randint(0, 255, (1, 3, 224, 224), dtype=torch.float32, device=device)   # this is the only difference vs previous code snippet
    eager = resize_uint8(image)
    compiled = compiled_resize_uint8(image)

    diff = eager.to(torch.float32) - compiled.to(torch.float32)
    print(
        seed,
        "mismatch:",
        (diff != 0).sum().item(),
        "max_abs:",
        diff.abs().max().item(),
    )

print(eager[0, 1, 36, 25], compiled[0, 1, 36, 25])
# (tensor(234.5000, device='cuda:0'), tensor(234.5000, device='cuda:0'))
```

EDIT:
You can try to get rid of this rounding problem at some point doing the following hack:
```python
def resize_uint8(x):
    x = x.to(torch.float32)
    y = F.resize(x, new_size, interpolation=interpolation, antialias=False)
    return (y + 1e-1).floor().to(torch.uint8)
```

**Comment by kaixuanliu** [comment]
@vfdev-5 Thx for digging into this, it makes sense. Would it be possible to make the compiled uint8 resize lowering use the same rounding mode as the eager implementation in pytorch side?

**Comment by vfdev-5** [comment]
Well, i do not know if we can do anything here to make outputs to match exactly. When I was saying it is a rounding problem, it is not exactly the difference of round op implementation. The implementations of the bilinear interpolate algorithm on cuda are not the same between eager and compiled, thus they lead to slightly different floating point results. The example above gives more precisely:
```python
print(eager[0, 1, 36, 25].item(), compiled[0, 1, 36, 25].item())
# 234.5 vs 234.50001525878906
```

Nearest mode should give exactly matching results. For bilinear, bicubic etc it is more complicated to get exactness across devices, dtypes.

Torchvision does the rounding here: https://github.com/pytorch/vision/blob/4252acbef3d7f37380860b0edb0f44424dcdfc0c/torchvision/transforms/v2/functional/_geometry.py#L359-L360

**Comment by eellison** [comment]
I was able to repro this with just aot_eager.

**Comment by williamwen42** [comment]
From claude:

## Investigation summary

  **Not a Dynamo/Inductor/cudagraphs bug** — it's the ATen `upsample_bilinear2d` decomposition rounding differently than the eager CUDA kernel. Expected float32 behavior, not a correctness regression.

  ### Layer isolation (backend bisect)

  | Backend | Adds over eager | uint8 mismatch (10 seeds) |
  |---|---|---|
  | `eager` | Dynamo capture only | **0** |
  | `aot_eager` | + AOTAutograd decomposition | **~25** (max abs 1) |

  Mismatch appears exactly when the decomposition is applied — no Inductor, no cudagraphs involved. So it's `_upsample_linear` in `torch/_decomp/decompositions.py` vs `aten/src/ATen/native/cuda/UpSampleBilinear2d.cu`.

  ### Why it can't be made bitwise-exact

  Both are equally accurate vs an fp64 reference (max err `0.0031297` for *both*; mean err `3.835e-4` eager vs `3.845e-4` decomp). They differ by ~1 ulp — two valid float32 roundings. Two causes:
  1. **Formula shape:** decomp `v1 + (v2-v1)*a` vs kernel `(1-a)*v1 + a*v2`. Scale/index/weights are otherwise identical.
  2. **FMA contraction:** nvcc fuses the kernel's `w0*v00 + w1*v01` into single-rounding FMAs; the decomposition emits discrete `mul`+`add` (two roundings). Aten ops can't reproduce this — I verified that aligning (1) alone does **not** fix it.

  Matches the earlier @vfdev-5 / @eellison assessment: only `nearest` is exact for compile.

  ### Recommendation

  **Close as known-limitation** (the `module: correctness (silent)` label already fits). Decomposed bilinear/trilinear is not bitwise-equivalent to the eager kernels by design.

  - A decomposition formula tweak (align to `(1-a)*v1 + a*v2`) is **not worth landing alone**: doesn't close the gap (FMA component remains), changes numerics for all bilinear/trilinear compile users (BC), and adds a multiply (the current 1-mul form is deliberate).
  - Genuinely closing the gap requires FMA-aware Inductor codegen for these decomps — a larger, separate effort.

  ### User workarounds

  - Compare with a **±1 LSB tolerance** (the error is bounded to 1).
  - `mode="nearest"` — bitwise-exact under compile.
  - `@torch.compiler.disable` the resize — runs the eager kernel, zero divergence.
  - @vfdev-5's bias hack: `(F.resize(x.float(), ...) + 0.1).floor().to(torch.uint8)`.

  ### Repro (no torchvision)

  ```python
  import torch, torch.nn.functional as F
  def resize_uint8(x):
      img = F.interpolate(x.float(), size=[48, 48], mode="bilinear",
                          align_corners=False, antialias=False)
      return img.round_().clamp_(0, 255).to(torch.uint8)

  for backend in ("eager", "aot_eager"):
      fn, total = torch.compile(resize_uint8, backend=backend), 0
      for seed in range(10):
          torch.compiler.reset(); torch.manual_seed(seed)
          x = torch.randint(0, 255, (1, 3, 224, 224), dtype=torch.uint8, device="cuda")
          total += (resize_uint8(x).to(torch.int16) - fn(x).to(torch.int16) != 0).sum().item()
      print(backend, total)   # eager: 0,  aot_eager: ~25

**Comment by kaixuanliu** [comment]
Thanks all! Close this issue now.
